root/tree/treeplayer/TDataFrameBenchmarks.cxx

```diff
@@ -0,0 +1,179 @@
+
+   gSystem->Unlink(fname.c_str());
+}
+BENCHMARK(TDF_NoOpLoopOnFile)->Unit(benchmark::kMicrosecond)->Arg(0)->Arg(1)->Arg(100)->Arg(100000);
+
+// measure time taken updating the custom column value and passing it to an action
+static void TDF_NoOpLoopOnCustomColumn(benchmark::State &state)
+{
+   for (auto _ : state)
+      TDataFrame(state.range(0)).Define("x", Zero).Foreach([](int) {}, {"x"});
+}
+BENCHMARK(TDF_NoOpLoopOnCustomColumn)->Unit(benchmark::kMicrosecond)->Arg(0)->Arg(1)->Arg(100)->Arg(100000);
+
+// measure time taken checking an upstream filter
+static void TDF_NoOpLoopWithFilter(benchmark::State &state)
+{
+  for (auto _ : state)
+     *TDataFrame(state.range(0)).Filter(Yes).Count();
+}
+BENCHMARK(TDF_NoOpLoopWithFilter)->Unit(benchmark::kMicrosecond)->Arg(0)->Arg(1)->Arg(100)->Arg(100000);
+
+static void TDF_BookHisto(benchmark::State &state)
+{
+   for (auto _ : state)
+      TDataFrame(0).Histo1D<ULong64_t>({"h", "h", 100, 0., 1.}, "tdfentry_");
+}
+BENCHMARK(TDF_BookHisto)->Unit(benchmark::kMicrosecond);
+
+// serves as baseline for the benchmarks that follow
+static void TDF_FillHistoNoTDF(benchmark::State &state)
+{
+   for (auto _ : state) {
+      auto nFills = state.range(0);
+      TH1D h("h", "h", nFills, 0., nFills);
+      for (auto i = 0; i < nFills; ++i)
+         h.Fill(i);
+   }
+}
+BENCHMARK(TDF_FillHistoNoTDF)->Unit(benchmark::kMicrosecond)->Arg(100)->Arg(100000);
+
+// serves as baseline for the benchmarks that follow
+static void TDF_FillHistoRangeDeductionNoTDF(benchmark::State &state)
+{
+   for (auto _ : state) {
+      auto nFills = state.range(0);
+      TH1D h("h", "h", nFills, 0., 0.);
+      for (auto i = 0; i < nFills; ++i)
+         h.Fill(i);
+   }
+}
+BENCHMARK(TDF_FillHistoRangeDeductionNoTDF)->Unit(benchmark::kMicrosecond)->Arg(100)->Arg(100000);
+
+// Fill a histogram from a single thread
+// Data is generated on the fly, not read from a file
+// The overhead of generating the data and passing it to the Histo action should be
+// the one measured by TDF_NoOpLoopOnCustomColumn
+static void TDF_FillHisto(benchmark::State &state)
+{
+   TDF::TH1DModel h("h", "h", state.range(0), 0., state.range(0));
+   for (auto _ : state)
+      *TDataFrame(state.range(0)).Histo1D<ULong64_t>(h, "tdfentry_");
+}
+BENCHMARK(TDF_FillHisto)->Unit(benchmark::kMicrosecond)->Arg(100)->Arg(100000);
+
+static void TDF_FillHistoRangeDeduction(benchmark::State &state)
+{
+   for (auto _ : state)
+      *TDataFrame(state.range(0)).Histo1D<ULong64_t>("tdfentry_");
+}
+BENCHMARK(TDF_FillHistoRangeDeduction)->Unit(benchmark::kMicrosecond)->Arg(100)->Arg(100000);
+
+// Fill a histogram from multiple-threads, merge partial results of each thread
+static void TDF_FillHistoMT(benchmark::State &state)
+{
+   TDF::TH1DModel h("h", "h", state.range(0), 0., state.range(0));
+   ROOT::EnableImplicitMT();
+   for (auto _ : state)
+      *TDataFrame(state.range(0)).Histo1D<ULong64_t>(h, "tdfentry_");
+   ROOT::DisableImplicitMT();
```

Comment:
You probably know better than me but last time I checked `ROOT::EnableImplicitMT()` calls `EnableThreadSafety` which seems to be irreversible.

We had a few issues with the TBufferMerger benchmarks. Namely, when we have a benchmark turning it on and then off. If one of the next benchmarks expects ThreadSafety to its default value we saw problems.

For these reasons, we decided that we should split the benchmarks requiring IMT into separate executables. The most stable way to do that was, moving the IMT benchmarks into a standalone executable and calling `ROOT::EnableImplicitMT()` in the initialization.

I'd suggest doing something similar here, because this will also give us a chance to benchmark builds of ROOT with `R__USE_IMT=Off`.

Reply:
To benchmark builds with `-Dimt=OFF` it's enough to add `#ifdef` guards to the relevant benchmarks (which I should have done in the first place, now fixed).

`EnableThreadSafety` initializes the global mutexes, which is irreversible, but `EnableImplicitMT` also regulates a global flag that tells `TDataFrame` whether to run in sequential or parallel mode, so `DisableImplicitMT` has the effect of deactivating implicit multi-threaded features in ROOT (although the global mutexes do remain activated).
So I think enabling and disabling IMT does the job here.

What kind of issues did you have with TBufferMerger exactly? Everything seems fine here, right?

Reply:
I think thread safety on/off has negligible impact on TDF runtimes. @dpiparo knows best, though.